// @sap-ux/adp-tooling/dist/writer/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrate = exports.generate = void 0;
const path_1 = require("path");
const mem_fs_1 = require("mem-fs");
const mem_fs_editor_1 = require("mem-fs-editor");
const ui5_config_1 = require("@sap-ux/ui5-config");
const options_1 = require("./options");
const tmplPath = (0, path_1.join)(__dirname, '../../templates/project');
/**
 * Set default values for optional properties.
 *
 * @param config configuration provided by the calling middleware
 * @returns enhanced configuration with default values
 */
function setDefaults(config) {
    var _a, _b, _c, _d, _e;
    var _f, _g, _h, _j;
    const configWithDefaults = {
        app: Object.assign({}, config.app),
        target: Object.assign({}, config.target),
        ui5: Object.assign({}, config.ui5),
        deploy: config.deploy ? Object.assign({}, config.deploy) : undefined,
        options: Object.assign({}, config.options)
    };
    (_a = (_f = configWithDefaults.app).title) !== null && _a !== void 0 ? _a : (_f.title = `Adaptation of ${config.app.reference}`);
    (_b = (_g = configWithDefaults.app).layer) !== null && _b !== void 0 ? _b : (_g.layer = 'CUSTOMER_BASE');
    (_c = configWithDefaults.package) !== null && _c !== void 0 ? _c : (configWithDefaults.package = config.package ? Object.assign({}, config.package) : {});
    (_d = (_h = configWithDefaults.package).name) !== null && _d !== void 0 ? _d : (_h.name = config.app.id.toLowerCase().replace(/\./g, '-'));
    (_e = (_j = configWithDefaults.package).description) !== null && _e !== void 0 ? _e : (_j.description = configWithDefaults.app.title);
    return configWithDefaults;
}
function writeUi5Yaml(basePath, config, fs) {
    return __awaiter(this, void 0, void 0, function* () {
        // ui5.yaml
        const ui5ConfigPath = (0, path_1.join)(basePath, 'ui5.yaml');
        const baseUi5ConfigContent = fs.read(ui5ConfigPath);
        const ui5Config = yield ui5_config_1.UI5Config.newInstance(baseUi5ConfigContent);
        (0, options_1.enhanceUI5Yaml)(ui5Config, config);
        fs.write(ui5ConfigPath, ui5Config.toString());
        // ui5-deploy.yaml
        if ((0, options_1.hasDeployConfig)(config)) {
            const ui5DeployConfig = yield ui5_config_1.UI5Config.newInstance(baseUi5ConfigContent);
            (0, options_1.enhanceUI5DeployYaml)(ui5DeployConfig, config);
            fs.write((0, path_1.join)(basePath, 'ui5-deploy.yaml'), ui5DeployConfig.toString());
        }
    });
}
/**
 * Writes the adp-project template to the mem-fs-editor instance.
 *
 * @param basePath - the base path
 * @param config - the writer configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
function generate(basePath, config, fs) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!fs) {
            fs = (0, mem_fs_editor_1.create)((0, mem_fs_1.create)());
        }
        const fullConfig = setDefaults(config);
        fs.copyTpl((0, path_1.join)(tmplPath, '**/*.*'), (0, path_1.join)(basePath), fullConfig, undefined, {
            globOptions: { dot: true },
            processDestinationPath: (filePath) => filePath.replace(/gitignore.tmpl/g, '.gitignore')
        });
        yield writeUi5Yaml(basePath, fullConfig, fs);
        return fs;
    });
}
exports.generate = generate;
/**
 * Writes the adp-project template to the mem-fs-editor instance during migration.
 *
 * @param basePath - the base path
 * @param config - the writer configuration
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
function migrate(basePath, config, fs) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!fs) {
            fs = (0, mem_fs_editor_1.create)((0, mem_fs_1.create)());
        }
        const fullConfig = setDefaults(config);
        // Copy the specified files to target project
        fs.copyTpl((0, path_1.join)(tmplPath, '**/ui5.yaml'), (0, path_1.join)(basePath), fullConfig, undefined, {
            globOptions: { dot: true }
        });
        fs.copyTpl((0, path_1.join)(tmplPath, '**/package.json'), (0, path_1.join)(basePath), fullConfig, undefined, {
            globOptions: { dot: true }
        });
        fs.copyTpl((0, path_1.join)(tmplPath, '**/gitignore.tmpl'), (0, path_1.join)(basePath), fullConfig, undefined, {
            globOptions: { dot: true },
            processDestinationPath: (filePath) => filePath.replace(/gitignore.tmpl/g, '.gitignore')
        });
        // delete .che folder
        if (fs.exists((0, path_1.join)(basePath, '.che/project.json'))) {
            fs.delete((0, path_1.join)(basePath, '.che/'));
        }
        // delete neo-app.json
        if (fs.exists((0, path_1.join)(basePath, 'neo-app.json'))) {
            fs.delete((0, path_1.join)(basePath, 'neo-app.json'));
        }
        yield writeUi5Yaml(basePath, fullConfig, fs);
        return fs;
    });
}
exports.migrate = migrate;
